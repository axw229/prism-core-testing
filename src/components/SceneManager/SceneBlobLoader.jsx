// @flow
// This component takes scene objects and loads the blob image
import React, { useEffect, useState } from 'react'
import ImageQueue from '../MergeCanvas/ImageQueue'
import {
  SYSTEM_ERROR, updateVariantsCollectionSurfaces
} from '../../store/actions/loadScenes'
import * as axios from 'axios'
import flattenDeep from 'lodash/flattenDeep'
import type { FlatScene, FlatVariant } from '../../shared/types/Scene'

type SceneBlobLoaderProps = {
  // @todo update this to be typed to the new scene structure
  scenes: FlatScene[],
  variants: FlatVariant[],
  initHandler?: Function,
  handleBlobsLoaded: Function,
  handleError: Function
}

const SceneBlobLoader = (props: SceneBlobLoaderProps) => {
  const { scenes, variants, initHandler, handleBlobsLoaded, handleError } = props
  const [blobUrls, setBlobUrls] = useState(null)
  const [loadCounter, setLoadCounter] = useState([])

  useEffect(() => {
    initHandler()
  }, [])

  useEffect(() => {
    if (scenes.length && !blobUrls) {
      let surfaces = variants.map(variant => variant.surfaces.map(surface => surface.mask))

      surfaces = flattenDeep(surfaces, 2)
      const blobPromises = surfaces.map((surface, i) => {
        return axios.get(surface, {
          responseType: 'blob'
        })
      })

      Promise.all(blobPromises)
        .then((resp) => {
          try {
            const _blobUrls = resp.map(r => URL.createObjectURL(r.data))
            setBlobUrls(_blobUrls)
          } catch {
            handleError({
              type: SYSTEM_ERROR,
              err: `Blob url not creation error`
            })
          }
        })
        .catch((err) => {
          handleError({
            type: SYSTEM_ERROR,
            err
          })
        })
    }
  }, [scenes, variants, blobUrls])

  useEffect(() => {
    if (blobUrls?.length) {
      if (loadCounter.length === blobUrls.length) {
        console.warn('Use event DOM references safely in the handleBlobsLoaded callback of SceneBlobLoader, do not hold on to them.')
        const variantsCollection = updateVariantsCollectionSurfaces(variants, blobUrls, 'surfaceBlobUrl')
        handleBlobsLoaded(variantsCollection)
      }
    }
  }, [blobUrls, loadCounter])

  // The index returned is the position of the original url, so that loaded urls can vbe organized irrespective of load order
  const loadImageAsBlobUrl = (e: SyntheticEvent, i: number) => {
    setLoadCounter([...loadCounter, i])
  }

  return (blobUrls?.length && blobUrls.length !== variants.length ? <ImageQueue dataUrls={blobUrls} addToQueue={loadImageAsBlobUrl} /> : null)
}

export default SceneBlobLoader
