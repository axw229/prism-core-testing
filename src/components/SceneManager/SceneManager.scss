:root {
  #{ getVarName('scenes.hitAreaOutlineColor') }: getVar('scenes.hitAreaOutlineColor');
}

.prism-scene-manager {
  $hit-area-outline-color: var(getVarName('scenes.hitAreaOutlineColor'));
  // TODO: Remove this once :root var declarations can be fixed in this build
  // hardcoding magenta as an outline color for now
  $hit-area-outline-color: magenta;

  $btn-active-color: $color-near-black;
  $btn-active-color-text: $color-white;
  $btn-focus-color: getVar('colors.primary');
  $btn-focus-color-text: $color-near-white;
  $btn-border-size: 2px;

  $icon-size: round($base * .7);
  $icon-spacing: round($base * .125);

  $self: &;

  @keyframes flashHitArea {
    0% {
      fill: rgba(white,0);
    }

    40% {
      fill: rgba(white,.1);
    }

    100% {
      fill: rgba(white,0);
    }
  }

  @keyframes pulseHitAreaOutline {
    0% {
      stroke: rgba($hit-area-outline-color, .5);
      stroke-width: 5px;
    }

    90%, 100% {
      stroke: rgba($hit-area-outline-color, 1);
      stroke-width: 10px;
    }
  }

  @keyframes fadeIn {
    0%, 25% {
      opacity: 0;
    }

    100% {
      opacity: 1;
    }
  }

  display: flex;
  flex-direction: column;
  font-size: $base;
  width: 100%;
  height: 100%;

  &__loader {
    color: getVar('colors.primary');
    left: 50%;
    position: absolute;
    top: 50%;
    transform: translateX(-50%) translateY(-50%);
  }

  &__block {
    align-items: center;
    display: flex;
    flex-direction: row;
    justify-content: center;

    // BEGIN TODO: REMOVE
    & > * {
      flex: 1 1;
    }
    // END TODO: REMOVE

    &--scenes {
      order: -1;
    }

    &--tabs {
      padding: 5px 0 0;
      justify-content: flex-start;
    }
  }

  &__flag {
    color: inherit;
    fill: currentColor;
    position: absolute;
    font-size: $icon-size;
    top: $icon-spacing;
    left: $icon-spacing;
    z-index: 1;
  }

  &__btn {
    float: left;
    margin: 0 0 0 $btn-border-size;
    max-width: 110px;
    padding: 0;
    position: relative;
    z-index: 1;
    color: $btn-active-color-text;

    &:first-child {
      margin-left: 0;
    }

    &::before {
      content: '';
      position: absolute;
      display: block;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
      z-index: 1;
    }

    &:hover, &:focus {
      color: $btn-focus-color-text;

      &::after {
        border-top-color: $btn-focus-color;
        border-left-color: $btn-focus-color;
      }

      &::before {
        box-shadow: inset 0 0 0 $btn-border-size $btn-focus-color;
      }
    }

    img {
      display: block;
      max-width: 100%;
    }

    &--active {
      &::after {
        content: '';
        top: 0;
        display: block;
        width: 0;
        height: 0;
        position: absolute;
        left: 0;
        border-top: $icon-size solid $btn-active-color;
        border-left: $icon-size solid $btn-active-color;
        border-bottom: $icon-size solid transparent;
        border-right: $icon-size solid transparent;
      }
    }

    &--disabled {
      opacity: .2;
    }
  }

  &__scene-wrapper {
    position: relative;
    overflow: hidden;
  }

  &__scene {
    background: black;
    clear: both;
    display: block;
    max-width: 100%;
    width: 100%;

    &--disabled {
      filter: saturate(0) brightness(.5) contrast(.75);
      pointer-events: none;
    }

    &__inner {
      height: 0;
      padding: 0;
      position: relative;
      width: 100%;
    }

    &__svg-defs {
      height: 0;
      position: absolute;
      width: 0;
      z-index: -1;
    }

    &__overlay,
    &__tint-wrapper,
    &__preview-wrapper,
    &__hit-wrapper {
      height: 100%;
      left: 0;
      position: absolute;
      top: 0;
      width: 100%;
    }

    &__natural {
      display: block;
      width: 100%;
      position: relative;
      height: auto;
    }

    &__preview-wrapper {
      pointer-events: none;

      &__preview {
        opacity: 0;
        transition: opacity .3s ease-in .0s;
        will-change: opacity;

        &--active {
          opacity: 1;
          transition: opacity .1s ease-out;
        }
      }
    }

    &__hit-area-wrapper,
    &__hit-area,
    &__surface {
      height: 100%;
      left: 0;
      pointer-events: none;
      position: absolute;
      top: 0;
      width: 100%;
    }

    &__hit-area {
      &__mask-loader {
        height: 0;
        position: absolute;
        width: 0;
        z-index: -1;
        visibility: hidden;
      }

      &__mask {
        cursor: pointer;
        fill: transparent;
        opacity: 0;
        pointer-events: all;
        stroke: rgba($hit-area-outline-color, .5);
        stroke-width: 5px;
        transition: opacity .4s ease-out;
        // don't you dare remove this line.
        will-change: opacity;

        &--hover,
        &:hover {
          animation-delay: .2s, 0s;
          animation-direction: normal, alternate;
          animation-duration: .4s, .6s;
          animation-iteration-count: 1, infinite;
          animation-name: flashHitArea, pulseHitAreaOutline;
          animation-timing-function: ease-in-out, ease-out;
          opacity: 1;
          transition: opacity .2s ease-in-out;
        }
      }
    }

    &__overlay {
      animation-name: fadeIn;
      animation-duration: .5s;
      animation-timing-function: ease-in-out;
      align-items: center;
      display: flex;
      justify-content: center;
      padding: .5em;
      background: pick-visible-color(getVar('colors.primary'), $color-black, $color-white);
      color: rgba(getVar('colors.primary'), $color-near-opacity);

      &__content {
        color: inherit;
        font-weight: bold;
        font-size: .8em;
        text-align: center;
      }

      &--loading {
        // same as default
      }

      &--error {
        background: rgba(getVar('colors.error'), $color-near-opacity);
        color: pick-visible-color(getVar('colors.error'), $color-black, $color-white);
      }

      &--message {
        background: rgba(getVar('colors.primary'), $color-near-opacity);
        color: pick-visible-color(getVar('colors.primary'), $color-black, $color-white);
      }
    }
  }
}