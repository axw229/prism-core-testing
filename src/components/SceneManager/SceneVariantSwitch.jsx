// @flow
import React, { PureComponent } from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { FormattedMessage } from 'react-intl'
import ReactGA from 'react-ga'
import includes from 'lodash/includes'

import { SCENE_VARIANTS } from 'constants/globals'

import './SceneVariantSwitch.scss'
type SwitchProps = {
  onChange: Function,
  currentVariant: string,
  sceneId: number
}

function SceneVariantSwitch () {}

SceneVariantSwitch.DayNight = class DayNight extends PureComponent<SwitchProps> {
  static name = 'day-night-toggle'
  static classes = {
    BASE: 'scene-variant-switch-day-night',
    CHECKBOX: 'visually-hidden',
    WRAPPER: 'scene-variant-switch-day-night__wrapper',
    DAY: 'scene-variant-switch-day-night__day',
    SWITCH: 'scene-variant-switch-day-night__switch',
    NIGHT: 'scene-variant-switch-day-night__night'
  }
  static isCompatible (variants: string[]): boolean {
    return includes(variants, SCENE_VARIANTS.DAY) && includes(variants, SCENE_VARIANTS.NIGHT)
  }

  constructor (props: SwitchProps) {
    super(props)

    this.handleChange = this.handleChange.bind(this)
    this.handleKeyDown = this.handleKeyDown.bind(this)
  }

  render () {
    const { currentVariant, sceneId } = this.props
    const checkboxName = `${DayNight.name}${sceneId}`
    const isDay = currentVariant === SCENE_VARIANTS.DAY

    return (
      <FormattedMessage id={isDay ? 'TO_NIGHT_VIEW' : 'TO_DAY_VIEW'}>
        {(txt: string) => (
          <label className={`${SceneVariantSwitch.DayNight.classes.BASE} ${!isDay ? `${SceneVariantSwitch.DayNight.classes.BASE}--night` : ''}`}
            onKeyDown={this.handleKeyDown}
            role='button'
            title={txt}
            aria-label={txt}
            htmlFor={checkboxName}
            tabIndex='0'>
            <input className={SceneVariantSwitch.DayNight.classes.CHECKBOX} type='checkbox' checked={!isDay} name={checkboxName} id={checkboxName} onChange={this.handleChange} />
            <div className={`${SceneVariantSwitch.DayNight.classes.WRAPPER} ${isDay ? `${SceneVariantSwitch.DayNight.classes.WRAPPER}--active` : ''}`}>
              <FontAwesomeIcon className={`${SceneVariantSwitch.DayNight.classes.DAY} ${!isDay ? `${SceneVariantSwitch.DayNight.classes.DAY}--active` : ''}`} icon={['fa', 'sun']} />
            </div>
            <div className={`${SceneVariantSwitch.DayNight.classes.WRAPPER} ${SceneVariantSwitch.DayNight.classes.WRAPPER}--night ${!isDay ? `${SceneVariantSwitch.DayNight.classes.WRAPPER}--active` : ''}`}>
              <FontAwesomeIcon className={`${SceneVariantSwitch.DayNight.classes.NIGHT} ${!isDay ? `${SceneVariantSwitch.DayNight.classes.NIGHT}--active` : ''}`} icon={['fa', 'moon-stars']} />
            </div>
          </label>
        )}
      </FormattedMessage>
    )
  }

  handleChange = function handleChange () {
    if (this.props.currentVariant === SCENE_VARIANTS.DAY) {
      this.props.onChange(SCENE_VARIANTS.NIGHT)
      ReactGA.event({
        category: 'Scene Manager',
        action: 'View Night Scene',
        label: 'View Night Scene'
      }, ['GAtrackerPRISM'])
    } else {
      this.props.onChange(SCENE_VARIANTS.DAY)
    }
  }

  handleKeyDown = function handleKeyDown (e: KeyboardEvent) {
    // if enter or space...
    if (e.keyCode === 32 || e.keyCode === 13) {
      e.stopPropagation()
      e.preventDefault()
      // ... act as though we've changed the checkbox
      this.handleChange()
    }
  }
}

export default SceneVariantSwitch
