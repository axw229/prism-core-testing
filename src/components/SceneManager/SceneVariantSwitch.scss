.scene-variant-switch-day-night {
  $base-font-size: $base; // 16px
  $icon-size: 1em;

  align-items: center;
  cursor: pointer;
  display: flex;
  float: right;
  font-size: $base-font-size;
  justify-content: center;
  max-width: 100%;
  position: absolute;
  right: 1rem;
  top: 1rem;
  z-index: 1;

  &__wrapper {
    align-items: center;
    background: rgba(41, 70, 147, 0.5);
    border-radius: 50%; 
    color: rgb(250, 232, 186);
    display: flex;
    filter: grayscale(100);
    height: $icon-size * 1.5;
    justify-content: center;
    opacity: 0.6;
    transform: scale(1);
    transition: filter 0.6s, opacity 0.6s, transform 0.6s;
    width: $icon-size * 1.5;
    will-change: opacity transform;

    &--active {
      background: rgba(41, 70, 147, 1);
      filter: grayscale(0);
      opacity: 1;
      transform: scale(1.35);
    }

    &--night {
      background: rgba(28, 28, 29, 0.9);
      color: rgb(208, 237, 255);
      margin-left: $icon-size * 0.5;

      &--active {
        background: rgba(28, 28, 29, 0.9);
      }
    }
  }

  &__day {
    height: $icon-size;
    width: $icon-size; 
  }

  &__night {
    height: $icon-size * 0.9;
    margin: 0 0 $icon-size * 0.1 $icon-size * 0.1;
    width: $icon-size * 0.9;
  }
}