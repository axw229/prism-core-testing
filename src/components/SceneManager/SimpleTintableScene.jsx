// @flow
import React, { useState } from 'react'
import { CSSTransition, TransitionGroup } from 'react-transition-group'
import type { Scene } from '../../shared/types/Scene'
import type { Color } from '../../shared/types/Colors'
import TintableSceneSurface from './TintableSceneSurface'
import TintableSceneSVGDefs from './TintableSceneSVGDefs'
import ensureFullyQualifiedAssetUrl from '../../shared/utils/ensureFullyQualifiedAssetUrl.util'
import { LiveMessage } from 'react-aria-live'
import { getFilterId, getMaskId } from '../../shared/utils/tintableSceneUtils'
import uniqueId from 'lodash/uniqueId'
import { baseClassName, getTintColorBySurface, transitionClassName } from './TintableScene'

type SimpleTintableSceneProps = {
  // eslint-disable-next-line react/no-unused-prop-types
  scene?: Scene,
  sceneType: string,
  sceneName: string,
  background: string,
  surfaceUrls: string[],
  surfaceIds: number[],
  highlights?: any[],
  shadows?: any[],
  colors: Color[],
  width: number,
  height: number,
  imageValueCurve?: any,
  isUsingWorkspace: boolean,
  activeColorId?: string
}

const reorderColors = (colors: Color[], activeColorId?: string) => {
  return colors.map((color, i) => {
    if (i === 0 && activeColorId) {
      return colors.find(c => c.id === activeColorId)
    } else if (color.id !== activeColorId) {
      return color
    } else {
      return null
    }
  }).filter(item => !!item)
}

const getTintColorBySurfaceAdapter = (isUsingWorkspace: boolean, surface: any, surfaceIndex: number, colors: Color[], activeColorId?: string) => {
  if (isUsingWorkspace) {
    const colours = reorderColors(colors, activeColorId)
    return colours[surfaceIndex]
  }

  // @todo, this is a stub it needs to be completely implemented to use -RS
  return getTintColorBySurface(surface)
}

const SimpleTintableScene = (props: SimpleTintableSceneProps) => {
  // @todo the scene prop is here for posterity, this component can be used as an adapter -RS
  const { sceneType, background, sceneName, width, height, imageValueCurve, surfaceUrls, isUsingWorkspace, surfaceIds, highlights, shadows, colors, activeColorId } = props
  const [instanceId] = useState(uniqueId('TS'))

  return (
    <>
      <div className={`${baseClassName}__tint-wrapper`}>
        <img className={`${baseClassName}__natural`} src={background} alt={sceneName} />
        <TransitionGroup className={`${transitionClassName}__colors`}>
          {surfaceUrls.map((surface: string, i) => {
            const highlight: ?string = highlights && surfaceUrls.length === highlights.length ? highlights[i] : null
            const shadow: ?string = shadows && surfaceUrls.length === shadows.length ? shadows[i] : null
            const tintColor: ?Color = getTintColorBySurfaceAdapter(isUsingWorkspace, surface, i, colors, activeColorId)
            if (tintColor) {
              return (
                <CSSTransition
                  key={`${surfaceIds[i]}_${tintColor.hex}`}
                  timeout={1}
                  mountOnEnter
                  classNames={`${transitionClassName}__colors__color-`} >
                  <TintableSceneSurface
                    type={sceneType}
                    image={background}
                    width={width}
                    height={height}
                    maskId={getMaskId(instanceId, surfaceIds[i], tintColor.hex)}
                    filterId={getFilterId(instanceId, surfaceIds[i], tintColor.hex)} >
                    <TintableSceneSVGDefs
                      type={sceneType}
                      width={width}
                      height={height}
                      highlightMap={highlight ? ensureFullyQualifiedAssetUrl(highlight) : void (0)}
                      shadowMap={shadow ? ensureFullyQualifiedAssetUrl(shadow) : void (0)}
                      filterId={getFilterId(instanceId, surfaceIds[i], tintColor.hex)}
                      filterColor={tintColor.hex}
                      filterImageValueCurve={imageValueCurve}
                      maskId={getMaskId(instanceId, surfaceIds[i], tintColor.hex)}
                      maskImage={surface} />
                  </TintableSceneSurface>
                </CSSTransition>
              )
            }
          })}
        </TransitionGroup>
      </div>

      {sceneName && (<LiveMessage message={`${sceneName} scene has been loaded`} aria-live='polite' />)}
    </>
  )
}

export default SimpleTintableScene
