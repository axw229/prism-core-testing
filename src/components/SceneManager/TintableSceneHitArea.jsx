// @flow
import React, { PureComponent } from 'react'
import { DRAG_TYPES } from 'constants/globals'
import { DropTarget } from 'react-dnd'

type Props = {
  connectDropTarget: Function,
  isOver: Boolean,
  id: string, // eslint-disable-line
  onDrop: Function, // eslint-disable-line
  onClick: Function, // eslint-disable-line
  color: string,
  svgSource: string
}

const TintableSceneHitAreaSpec = {
  drop (props: Props, monitor) {
    const droppedItem = monitor.getItem()

    if (droppedItem && droppedItem.color && droppedItem.color.hex) {
      props.onDrop(props.id, droppedItem.color.hex)
    }
  }
}

function collect (connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver(),
    canDrop: monitor.canDrop()
  }
}

class TintableSceneHitArea extends PureComponent<Props> {
  static defaultProps = {}
  static baseClass = 'prism-scene-manager__scene__hit-area'
  static baseClassMask = `${TintableSceneHitArea.baseClass}__mask`

  constructor (props) {
    super(props)

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick = function handleClick (e: any) {
    this.props.onClick(this.props.id)
  }

  render () {
    const { connectDropTarget, isOver, color, svgSource } = this.props
    return connectDropTarget && connectDropTarget(
      <svg className={TintableSceneHitArea.baseClass} style={{ fill: color }}>
        <use className={`${TintableSceneHitArea.baseClassMask} ${isOver ? `${TintableSceneHitArea.baseClassMask}--hover` : ''}`}
          xlinkHref={`${svgSource}#mask`}
          onClick={this.handleClick} />
      </svg>
    )
  }
}

export default DropTarget(DRAG_TYPES.SWATCH, TintableSceneHitAreaSpec, collect)(TintableSceneHitArea)
