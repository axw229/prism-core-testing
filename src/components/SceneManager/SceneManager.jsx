// @flow
import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import find from 'lodash/find'
import includes from 'lodash/includes'
import memoizee from 'memoizee'
import ReactGA from 'react-ga'

import { SCENE_TYPES, SCENE_VARIANTS } from 'constants/globals'
import { loadScenes, paintSceneSurface, activateScene, deactivateScene, changeSceneVariant } from '../../store/actions/scenes'
import TintableScene from './TintableScene'
import SceneVariantSwitch from './SceneVariantSwitch'
import ImagePreloader from '../../helpers/ImagePreloader'
import { ensureFullyQualifiedAssetUrl } from '../../shared/helpers/DataUtils'
import CircleLoader from '../Loaders/CircleLoader/CircleLoader'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import type { Color } from '../../shared/types/Colors'
import type { Scene, SceneStatus, Surface, Variant } from '../../shared/types/Scene'

import './SceneManager.scss'

type Props = {
  scenes: Scene[],
  sceneStatus: SceneStatus[],
  type: string,
  activeScenes: Array<number | string>,
  maxActiveScenes: number,
  loadScenes: Function,
  activateScene: Function,
  deactivateScene: Function,
  paintSceneSurface: Function,
  changeSceneVariant: Function,
  loadingScenes: boolean,
  activeColor: Color | void,
  mainColor: Color | void,
  previewColor: Color | void,
  interactive: boolean
}

type State = {
  currentSceneIndex: number
}

class SceneManager extends PureComponent<Props, State> {
  static baseClass = 'prism-scene-manager'

  static defaultProps = {
    maxActiveScenes: 2,
    type: SCENE_TYPES.ROOM,
    interactive: true
  }

  state = {
    currentSceneIndex: 0
  }

  constructor (props) {
    super(props)

    this.handleColorUpdate = this.handleColorUpdate.bind(this)
    this.handleClickSceneToggle = this.handleClickSceneToggle.bind(this)
    this.changeVariant = this.changeVariant.bind(this)
  }

  componentDidMount () {
    this.props.loadScenes(this.props.type)
  }

  handleColorUpdate = function handleColorUpdate (sceneId, surfaceId, color: Color) {
    this.props.paintSceneSurface(sceneId, surfaceId, color)
  }

  handleClickSceneToggle = function handleClickSceneToggle (id) {
    const { activeScenes } = this.props

    // if this scene is active...
    if (includes(activeScenes, id)) {
      this.deactivateScene(id)
    } else {
      // ... otherwise activate this scene
      this.activateScene(id)
      ReactGA.event({
        category: 'Scene Manager',
        action: 'Toggle Active Scene',
        label: 'Toggle Active Scene'
      })
    }
  }

  activateScene (id) {
    const { activeScenes, maxActiveScenes, activateScene, deactivateScene } = this.props

    // if active scenes exceed or match max active scenes...
    if (activeScenes.length >= parseInt(maxActiveScenes, 10)) {
      // compare prev and current scenes, continut to eliminate oldest until max active is all that remains
      deactivateScene(activeScenes.slice(0, activeScenes.length - maxActiveScenes + 1))
    }

    activateScene(id)
  }

  deactivateScene (id) {
    const { activeScenes, deactivateScene } = this.props

    if (activeScenes.length === 1) {
      // don't let the user deactivate the last scene
      return
    }
    // ... deactivate it
    deactivateScene(id)
  }

  changeVariant = memoizee(function changeVariant (sceneId: number) {
    const { changeSceneVariant } = this.props

    return function _changeVariant (variant: string) {
      changeSceneVariant(sceneId, variant)
    }
  }, { primitive: true, length: 1 })

  render () {
    const { scenes, sceneStatus, loadingScenes, activeColor, previewColor, mainColor, activeScenes, type, interactive } = this.props

    if (loadingScenes) {
      return <CircleLoader className={`${SceneManager.baseClass}__loader`} />
    }

    return (
      <div className={SceneManager.baseClass}>
        <div className={`${SceneManager.baseClass}__block ${SceneManager.baseClass}__block--tabs`}>
          {/* POC scene-switching buttons to demonstrate performance */}
          {scenes.map((scene, index) => {
            const sceneId = scene.id

            const status: SceneStatus = find(sceneStatus, { 'id': sceneId })
            const sceneVariant: Variant = find(scene.variants, { 'variant_name': status.variant })
            const surfaces: Surface[] = sceneVariant.surfaces
            const activeMarker = includes(activeScenes, scene.id)
              ? <FontAwesomeIcon icon={['fa', 'check']} className={`${SceneManager.baseClass}__flag`} />
              : null

            return (
              // TODO: Convert these to labels around checkboxes since that's how they're being used
              <button key={sceneId}
                onClick={() => this.handleClickSceneToggle(scene.id)}
                className={`${SceneManager.baseClass}__btn ${includes(activeScenes, scene.id) ? `${SceneManager.baseClass}__btn--active` : ''}`}
                type='button'>

                <span className='visually-hidden'>{sceneVariant.name}</span>
                <div role='presentation'>
                  <ImagePreloader
                    el={TintableScene}
                    preload={[
                      sceneVariant.image,
                      surfaces.map(surface => surface.shadows),
                      surfaces.map(surface => surface.mask),
                      surfaces.map(surface => surface.hitArea),
                      surfaces.map(surface => surface.highlights)
                    ]}
                    interactive={false}
                    width={scene.width}
                    height={scene.height}
                    type={type}
                    sceneId={sceneId}
                    background={ensureFullyQualifiedAssetUrl(sceneVariant.image)}
                    clickToPaintColor={activeColor}
                    onUpdateColor={this.handleColorUpdate}
                    previewColor={previewColor}
                    mainColor={mainColor}
                    surfaceStatus={status.surfaces}
                    surfaces={surfaces}
                  />
                </div>
                {activeMarker}
              </button>
            )
          })}
        </div>

        <div className={`${SceneManager.baseClass}__block ${SceneManager.baseClass}__block--scenes`}>
          {activeScenes.map((sceneId, index) => {
            const scene: Scene = scenes.filter(scene => (scene.id === sceneId))[0]

            if (!scene) {
              return null
            }

            const status: SceneStatus = find(sceneStatus, { 'id': sceneId })
            const sceneVariant: Variant = find(scene.variants, { 'variant_name': status.variant })
            const surfaces: Surface[] = sceneVariant.surfaces

            let variantSwitch = null

            // if we have more than one variant for this scene...
            if (scene.variant_names.length > 1) {
              // if we have day and night variants...
              if (SceneVariantSwitch.DayNight.isCompatible(scene.variant_names)) {
                // ... then create a day/night variant switch
                variantSwitch = <SceneVariantSwitch.DayNight currentVariant={status.variant} variants={[SCENE_VARIANTS.DAY, SCENE_VARIANTS.NIGHT]} onChange={this.changeVariant(sceneId)} />
              }
            }

            return (
              <div className={`${SceneManager.baseClass}__scene-wrapper`} key={sceneId}>
                <ImagePreloader
                  el={TintableScene}
                  preload={[
                    sceneVariant.image,
                    surfaces.map(surface => surface.mask),
                    surfaces.map(surface => surface.shadows),
                    surfaces.map(surface => surface.hitArea),
                    surfaces.map(surface => surface.highlights)
                  ]}
                  width={scene.width}
                  height={scene.height}
                  interactive={interactive}
                  type={type}
                  sceneId={sceneId}
                  background={ensureFullyQualifiedAssetUrl(sceneVariant.image)}
                  clickToPaintColor={activeColor}
                  onUpdateColor={this.handleColorUpdate}
                  previewColor={previewColor}
                  mainColor={mainColor}
                  surfaceStatus={status.surfaces}
                  surfaces={surfaces}
                  imageValueCurve={sceneVariant.normalizedImageValueCurve}
                  sceneName={sceneVariant.name}
                />
                {variantSwitch}
              </div>
            )
          })}
        </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => {
  let activeColor = void (0)
  let previewColor = void (0)

  if (state.lp.activeColor) {
    activeColor = state.lp.activeColor
  }

  if (state.lp.activePreviewColor) {
    previewColor = state.lp.activePreviewColor
  }

  return {
    scenes: state.scenes.sceneCollection[state.scenes.type],
    sceneStatus: state.scenes.sceneStatus[state.scenes.type],
    numScenes: state.scenes.numScenes,
    activeScenes: state.scenes.activeScenes,
    loadingScenes: state.scenes.loadingScenes,
    activeColor: activeColor,
    previewColor: previewColor
    // NOTE: uncommenting this will sync scene type with redux data
    // we may not want that in case there are multiple instances with different scene collections running at once
    // leaving here for posterity
    // type: state.scenes.type
  }
}

const mapDispatchToProps = (dispatch: Function) => {
  return {
    loadScenes: (type: string) => {
      dispatch(loadScenes(type))
    },
    paintSceneSurface: (sceneId, surfaceId, color) => {
      dispatch(paintSceneSurface(sceneId, surfaceId, color))
    },
    activateScene: (sceneId) => {
      dispatch(activateScene(sceneId))
    },
    deactivateScene: (sceneId) => {
      dispatch(deactivateScene(sceneId))
    },
    changeSceneVariant: (sceneId: number, variant: string) => {
      dispatch(changeSceneVariant(sceneId, variant))
    }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(SceneManager)
