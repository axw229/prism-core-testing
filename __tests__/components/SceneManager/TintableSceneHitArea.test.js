import React from 'react'
import { shallow } from 'enzyme'
import { TintableSceneHitArea } from 'src/components/SceneManager/TintableSceneHitArea'

const defaultProps = {
  connectDropTarget: jest.fn((el) => el),
  onDrop: jest.fn(),
  onClick: jest.fn(),
  onOver: jest.fn(),
  onOut: jest.fn(),
  onLoadingSuccess: jest.fn(),
  onLoadingError: jest.fn()
}

const createHitArea = (props = {}) => {
  return shallow(<TintableSceneHitArea {...defaultProps} {...props} />)
}

// snapshot testing
describe('Snapshot Testing', () => {
  it('component should match snapshot', () => {
    const wrapper = createHitArea()
    expect(wrapper).toMatchSnapshot()
  })
})

// rendering test
describe('testing Tintable Scene hit area', () => {
  it('SVG should rendering correctly', () => {
    const wrapper = createHitArea()
    expect(wrapper.find('InlineSVG').exists()).toBe(true)
    expect(wrapper.find('svg').exists()).toBe(true)
  })
})

// event test
describe('event testing', () => {
  it('click function should be called correctly', () => {
    const wrapper = createHitArea()
    const click = jest.fn()
    wrapper.instance().handleClick = click
    wrapper.instance().forceUpdate()
    wrapper.find('use').simulate('click')
    expect(click).toHaveBeenCalled()
  })
})
