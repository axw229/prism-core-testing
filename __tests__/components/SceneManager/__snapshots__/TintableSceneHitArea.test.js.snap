// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing component should match snapshot 1`] = `
<div
  className="prism-scene-manager__scene__hit-area-wrapper"
>
  <InlineSVG
    baseURL=""
    cacheGetRequests={false}
    cacheRequests={true}
    className="prism-scene-manager__scene__hit-area__mask-loader"
    onError={[Function]}
    onLoad={[Function]}
    src="localhostundefined"
    supportTest={[Function]}
    uniqueHash="TSHA1"
    uniquifyIDs={true}
    wrapper={[Function]}
  />
  <svg
    className="prism-scene-manager__scene__hit-area"
  >
    <use
      className="prism-scene-manager__scene__hit-area__mask "
      onClick={[Function]}
      xlinkHref="#mask___TSHA1"
    />
  </svg>
</div>
`;
